kymatio/scattering2d/backend/backend_torch.py

```diff
@@ -174,31 +178,39 @@ def cdgmm(A, B, inplace=False):
             C[b, c, m, n, :] = A[b, c, m, n, :] * B[m, n, :]
     """
     A, B = A.contiguous(), B.contiguous()
-    if A.size()[-3:] != B.size():
+    if A.size()[-3:-1] != B.size()[-3:-1]:
         raise RuntimeError('The filters are not compatible for multiplication!')
 
-    if not iscomplex(A) or not iscomplex(B):
-        raise TypeError('The input, filter and output should be complex')
-
-    if B.ndimension() != 3:
-        raise RuntimeError('The filters must be simply a complex array!')
+    if not iscomplex(A):
+        raise TypeError('The input must be complex')
 
-    if type(A) is not type(B):
-        raise RuntimeError('A and B should be same type!')
+    if not iscomplex(B) and not isreal(B):
+        raise TypeError('The filter must be complex or real')
 
+    if B.ndimension() != 3:
+        raise RuntimeError('The filter must be a 3-tensor')
 
-    C = A.new(A.size())
-
-    A_r = A[..., 0].contiguous().view(-1, A.size(-2)*A.size(-3))
-    A_i = A[..., 1].contiguous().view(-1, A.size(-2)*A.size(-3))
+    if A.dtype is not B.dtype:
+        raise RuntimeError('A and B must be of the same dtype')
 
-    B_r = B[...,0].contiguous().view(B.size(-2)*B.size(-3)).unsqueeze(0).expand_as(A_i)
-    B_i = B[..., 1].contiguous().view(B.size(-2)*B.size(-3)).unsqueeze(0).expand_as(A_r)
+    if A.device != B.device:
+        raise RuntimeError('A and B must be on the same device')
 
-    C[..., 0].view(-1, C.size(-2)*C.size(-3))[:] = A_r * B_r - A_i * B_i
-    C[..., 1].view(-1, C.size(-2)*C.size(-3))[:] = A_r * B_i + A_i * B_r
+    if isreal(B):
+        if inplace:
+            return A.mul_(B)
+        else:
+            return A * B
```

Comment:
In this concise formulation there could be a lesson for cleaning up the code below :)

(not for now, for another time, to be benchmarked because of the contiguity issues)

Reply:
Definitely. I found the code below barely readable, but now it's unused so I had no strong motivation to rewrite it.